demo-app/src/tx_hooks.rs

```diff
@@ -0,0 +1,63 @@
+use crate::tx_verifier::{Transaction, VerifiedTx};
+use sov_modules_api::Context;
+use sov_state::WorkingSet;
+
+/// TxHooks allows injecting custom logic into a transaction processing pipeline.
+pub(crate) trait TxHooks {
+    type Context: Context;
+
+    /// pre_dispatch_tx_hook runs just before a transaction is dispatched to an appropriate module.
+    fn pre_dispatch_tx_hook(
+        &mut self,
+        tx: Transaction<Self::Context>,
+    ) -> anyhow::Result<VerifiedTx<Self::Context>>;
+
+    /// post_dispatch_tx_hook runs after the tx is dispatched to an appropriate module.
+    fn post_dispatch_tx_hook(&mut self, tx: VerifiedTx<Self::Context>);
+}
+
+pub(crate) struct DemoAppTxHooks<C: Context> {
+    accounts_hooks: accounts::hooks::Hooks<C>,
+}
+
+impl<C: Context> DemoAppTxHooks<C> {
+    pub fn new(working_set: WorkingSet<C::Storage>) -> Self {
+        Self {
+            accounts_hooks: accounts::hooks::Hooks::<C>::new(working_set),
+        }
+    }
+}
+
+impl<C: Context> TxHooks for DemoAppTxHooks<C> {
+    type Context = C;
+
+    fn pre_dispatch_tx_hook(
+        &mut self,
+        tx: Transaction<Self::Context>,
+    ) -> anyhow::Result<VerifiedTx<Self::Context>> {
+        let acc = self
+            .accounts_hooks
+            .get_or_create_default_account(tx.pub_key.clone())?;
+
+        let tx_nonce = tx.nonce;
+        let acc_nonce = acc.nonce;
+        anyhow::ensure!(
+            acc_nonce == tx_nonce,
+            "Tx bad nonce, expected: {acc_nonce}, but found: {tx_nonce}",
+        );
+
+        Ok(VerifiedTx {
+            pub_key: tx.pub_key,
+            sender: acc.addr,
+            runtime_msg: tx.runtime_msg,
+        })
```

Comment:
non-blocking: I think this logic is going to be used in basically every `hooks` for a chain that uses the accounts module. Maybe we can abstract it into a function?